### PR TITLE
Build Gates C++ Solution

### PR DESCRIPTION
**Pull request by maggie-j-liu**
I added a C++ solution to Build Gates (January 2016 Silver). 

I also had a question about the [official solution](http://www.usaco.org/current/data/sol_gates_silver_jan16.html), they only made the farm 2005 by 2005 and started at (1002, 1002), but they also doubled the length of the fence. Wouldn't this go out of bounds for an input of 1000 Ns?

### CODE REVIEW

**Comment by frodakcin** [comment]
Yea, I think it does go out of bounds.

**Review by frodakcin — APPROVED** [review]
Clean -- should we possibly point out the potential error in the official analysis?

**Comment by maggie-j-liu** [comment]
Sure, I'll add a note about the official analysis in the solution.

File: solutions/USACO jan16-build-gates.mdx [hunk 1]
```diff
@@ -0,0 +1,89 @@
+---
+id: usaco-596
+source: USACO Silver 2016 January
+title: Build Gates
+author: Maggie Liu
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_gates_silver_jan16.html)
+
+The official solution stores the farm in a $2005 \times 2005$ boolean array with the origin at $(1002, 1002)$, but this is not sufficient. Since each fence segment is doubled, the maximum distance in each direction could be $2000$ units, causing an input such as $1000$ $\texttt{N}$s to go out of bounds. To not go out of bounds, an array of size $4003 \times 4003$ should be used: $1$ unit for the origin position, $2000$ units in each direction from the origin and $1$ more unit in each direction to ensure that the farm includes area beyond the fence.
```
Comment (HarryWangATX) [inline solutions/USACO jan16-build-gates.mdx:10]: Wrap in warning???

**Comment by HarryWangATX** [comment]
Everything else seems good!

File: solutions/USACO jan16-build-gates.mdx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+source: USACO Silver 2016 January
+title: Build Gates
+author: Maggie Liu
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_gates_silver_jan16.html)
+
+<Warning>
+
+The official solution stores the farm in a $2005 \times 2005$ boolean array with the origin at $(1002, 1002)$, but this is not sufficient. Since each fence segment is doubled, the maximum distance in each direction could be $2000$ units, causing an input such as $1000$ $\texttt{N}$s to go out of bounds. To not go out of bounds, an array of size $4003 \times 4003$ should be used: $1$ unit for the origin position, $2000$ units in each direction from the origin and $1$ more unit in each direction to ensure that the farm includes area beyond the fence.
```
Comment (dongliu0426) [inline solutions/USACO jan16-build-gates.mdx:12]: use `\cdot` instead of `\times`

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

File: solutions/USACO jan16-build-gates.mdx (hunk 2 above)
Comment (maggie-j-liu) [reply solutions/USACO jan16-build-gates.mdx:12]: But the `\times` here is equivalent to "by" in specifying the dimensions of the array, like `N \times M` [here](https://usaco.guide/silver/ff#implementation). Should it be changed to `\cdot`?

Comment (HarryWangATX) [reply solutions/USACO jan16-build-gates.mdx:12]: i'm thinking times as well???

**Review by HarryWangATX — APPROVED** [review]

File: solutions/USACO jan16-build-gates.mdx (hunk 2 above)
Comment (dongliu0426) [reply solutions/USACO jan16-build-gates.mdx:12]: oops you're right.

**Review by dongliu0426 — APPROVED** [review]

**Pull request merged**